packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -322,22 +397,120 @@ class TimelineEvent {
   TimelineEvent parent;
   List<TimelineEvent> children = <TimelineEvent>[];
 
-  int get duration => endTime - startTime;
-
-  bool get wellFormed => startTime != null && duration != null;
+  int get duration => (endTime != null) ? endTime - startTime : null;
 
   bool get isCpuEvent => type == TimelineEventType.cpu;
   bool get isGpuEvent => type == TimelineEventType.gpu;
 
+  bool get isPipelineProduceFlow => _hasChild('Engine::BeginFrame');
+  bool get isPipelineConsumeFlow => _hasChild('PipelineConsume');
+
+  /// Whether there is a child with the given name [childName] is contained
+  /// somewhere in the subtree [children].
+  bool _hasChild(String childName) {
+    bool childFound = false;
+
+    void findChild(TimelineEvent event, String childName) {
+      if (event.name.contains(childName)) {
+        childFound = true;
+      }
+      if (childFound) return;
+      for (TimelineEvent e in event.children) {
+        findChild(e, childName);
+      }
+    }
+
+    findChild(this, childName);
+    return childFound;
+  }
+
+  /// Adds a child event.
+  ///
+  /// If we know the child's location (knownLocation == true)
+  /// add the child directly via [_addChild]. If we do not know the child's
+  /// location, find it's proper place in the subtree and add the child there.
+  void addChild(TimelineEvent child, {bool findChildLocation = false}) {
+    bool newChildAdded = false;
```

Comment:
use return values from functions to eliminate the newChildAdded variable from this method. That will make the logic flow clearer.

Reply:
Found a way to remove newChildAdded without adding return values.